lnwallet/channel.go

```diff
@@ -7621,3 +7703,95 @@ func (lc *LightningChannel) unsignedLocalUpdates(remoteMessageIndex,
 
 	return localPeerUpdates
 }
+
+// GenMusigNonces generates the verification nonce to start off a new musig2
+// channel session.
+func (lc *LightningChannel) GenMusigNonces() (*musig2.Nonces, error) {
+	lc.RLock()
+	defer lc.RUnlock()
+
+	var err error
+	lc.pendingVerificationNonce, err = NewMusigVerificationNonce(
+		lc.channelState.LocalChanCfg.MultiSigKey.PubKey,
+		lc.currentHeight, lc.channelState.RevocationProducer,
+		false,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return lc.pendingVerificationNonce, nil
+}
+
+// NewMusigVerificationNonce generates the local or verification nonce for
+// another musig2 session. In order to permit our implementation to not have to
+// write any secret nonce state to disk, we'll use the _next_ shachain
```

Comment:
Is there really no way in which that next shachain pre-image can be used in different signing contexts?

Reply:
Updated this to use a _distinct_ sha chain revocation root now. This is derived from the current root (note that the root is only revealed to the remote party once all 2^48-1 states are known to them) via an `hmac` invocation. 

> Is there really no way in which that next shachain pre-image can be used in different signing contexts?

So this is the nonce we use to sign our own commitments. We only sign our commitment when we go to force close. We only ever force close once (at that point we typically only have a single valid commitment). The nonce we generate is bound to the height that we're closing for, and by definition, we can only ever close once. On top of all that, we can only sign the _same_ message for a given height, as the message is already bound by the signature of the remote party (we only have one state per height).